### PR TITLE
Router class implementation

### PR DESCRIPTION
**Pull request by ismailAde**
The router class gets user request through the url and breaks down the
url to determine which controller and what action is to be called as
well as parameters to pass along.
#Changes were made on a windows server

### CODE REVIEW

File: app/init.php [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+defined('ROOT_PATH')? null: define('ROOT_PATH', getcwd());
+defined('APP_PATH')? null: define('APP_PATH', __DIR__);
+defined('DS')? null: define('DS', DIRECTORY_SEPARATOR);
+
+require_once(APP_PATH.DS.'core/Router.php');
+require_once(APP_PATH.DS.'core/Bootstrap.php');
+require_once(ROOT_PATH.DS.'vendor/autoload.php');
```
Comment (perfectmak) [inline app/init.php]: Please remove this file and use composer class loader. Look into the composer.json file for details of the class map. We are using psr-4

File: index.php [hunk 2]
```diff
@@ -1,7 +1,9 @@
 <?php
+require_once('app/init.php');
+//require(__DIR__ . '/vendor/autoload.php');
```
Comment (perfectmak) [inline index.php]: uncomment this and remove the app/init.php require.

File: index.php [hunk 3]
```diff
@@ -1,7 +1,9 @@
...
 
-use App\Core\Dispatcher;
+use App\Core\Bootstrap as Bootstrap;
+use App\Core\Router as Router;
```
Comment (perfectmak) [inline index.php]: You do not need the `as` statements as it is redundant here, so please remove them too

File: .htaccess [hunk 4]
```diff
@@ -0,0 +1,9 @@
+options -MultiViews
+RewriteEngine on
+
+RewriteBase /voucher-service-develop
```
Comment (perfectmak) [inline .htaccess]: RewriteBase here would prevent this codebase from being portable so remove it

File: app/Core/Router.php [hunk 5]
```diff
@@ -27,6 +33,54 @@ private function __construct() {
      * @return Router
      */
     public static function getInstance() : Router {
-        return self::$routerInstance ?? new Router();
+        return $this->$routerInstance ?? new Router();
+    }
+
+    public function parseUrl($url = ''){
```
Comment (perfectmak) [inline app/Core/Router.php]: This function has no business here. Remove it and check out my last comment for better explanation of why.
And the same goes for the `getController()`, `getAction()` and `getParams()`

**Comment by perfectmak** [comment]
@ismailAde This is good 👍, but the job of the router is to register routes and the callbacks that would be invoked when such routes in invoked.

So it has no business determining which controller should be used based on the url or even breaking down the url. Another class would be made to do that if required.

So it shouldn't have controller or actions information.
It should basically expose an interface that allows registering of routes and the callbacks. And also allows retrieving the callback based on the routes. So the Router should be like a Map class.

**Comment by perfectmak** [comment]
Also @ismailAde rename the controllers directory to conform with ps-4 standards.

**Comment by ismailAde** [comment]
Kindly go through the project now @perfectmak .

File: app/Core/Router.php [hunk 6]
```diff
@@ -29,4 +42,74 @@ private function __construct() {
...
     }
+
+    /**
+     *  Get Request on url and action
+     *
+     * @param string url
+     * @param string action
+     */
+    public function get($url, $action){
+        $this->addRoute('Get', $url, $action);
```
Comment (perfectmak) [inline app/Core/Router.php]: Make the 'Get' string a class constant and reference it here, also do the same for the other types.

File: app/routes.php [hunk 7]
```diff
@@ -6,3 +6,5 @@
  * Date: 12/21/16
```
Comment (perfectmak) [inline app/routes.php]: Please delete this file and this folder from this PR

File: index.php [hunk 8]
```diff
@@ -1,7 +1,5 @@
 <?php
-
 use App\Core\Dispatcher;
```
Comment (perfectmak) [inline index.php]: Remove this use statement, it is redundant

**Comment by perfectmak** [comment]
@ismailAde Great job. So kindly go through the comments and revert as soon as possible.

**Comment by perfectmak** [comment]
@ismailAde Great job. So kindly go through the comments and revert as soon as possible.

**Comment by ismailAde** [comment]
Noted, Will work on that now

File: app/routes.php (hunk 7 above)
Comment (ismailAde) [reply app/routes.php]: The routes.php file will be deleted now. But its in the App directory, Should the folder be deleted as well?

**Comment by ismailAde** [comment]
Done @perfectmak

File: app/Core/Router.php [hunk 9]
```diff
@@ -29,4 +51,74 @@ private function __construct() {
...
+    }
+
+    /**
+     * Adds newly registered route to routes array
+     *
+     * @param string $request_type
+     * @param string url
+     * @param string action
+     */
+    protected function addRoute($request_type='Get', $url, $action){
```
Comment (perfectmak) [inline app/Core/Router.php]: Does this line work? I mean PHP doesn't allow default parameters as the first parameter?
Which brings me to the question, have you tested this class to work?

Comment (ismailAde) [reply app/Core/Router.php]: It skipped my mind the initial commit was not tested. However the line works. I tested and made another commit.

File: app/Core/Router.php [hunk 10]
```diff
@@ -29,4 +51,83 @@ private function __construct() {
...
+    }
+
+    /**
+     * Adds newly registered route to routes array
+     *
+     * @param string $request_type
+     * @param string url
+     * @param string action
+     */
+    protected function addRoute($request_type=self::GET, $url, $action){
```
Comment (perfectmak) [inline app/Core/Router.php]: This line would not work, the default argument would throw an error.
Check here: http://php.net/manual/en/functions.arguments.php for more infor. Look under incorrect usage of default function argument.

**Pull request merged**